Added EventParticipationController and UserMapper

SoslanDzgoev opened the pull request:
- Implemented EventParticipationController with methods to register and unregister participants, and to get participants list and count
- Added UserMapper interface to map between User and UserDTO objects

SoslanDzgoev committed:
Added EventParticipationController and UserMapper

- Implemented EventParticipationController with methods to register and unregister participants, and to get participants list and count
- Added UserMapper interface to map between User and UserDTO objects

efr21 commented on src/main/java/school/faang/user_service/service/event/EventParticipationService.java:
```diff
@@ -0,0 +1,59 @@
+    private EventParticipationRepository eventParticipationRepository;
+
+    public void registerParticipant(Long eventId, Long userId) {
+        List<User> participants = eventParticipationRepository.findAllParticipantsByEventId(eventId);
```
Возможно, стоит рассмотреть такой вариант: проверять, не зарегистрирован ли уже пользователь на событии, на уровне repository (exists). Таким образом, не придется проходить по списку всех участиков события, который может быть достаточно большим (размер этого списка мы не можем предсказать).

SoslanDzgoev replied:
Согласен, но в репозитории не было метода для получения участников мероприятия по id

efr21 commented on src/main/java/school/faang/user_service/service/event/EventParticipationService.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public void unregisterParticipant(Long eventId, Long userId) {
+        List<User> participants = eventParticipationRepository.findAllParticipantsByEventId(eventId);
```
Здесь тоже, возможно, стоит сделать через exists.

efr21 commented on src/main/java/school/faang/user_service/service/event/EventParticipationService.java:
```diff
@@ -0,0 +1,59 @@
+        List<User> participants = eventParticipationRepository.findAllParticipantsByEventId(eventId);
+
+        if (participants.isEmpty()) {
+            return new ArrayList<>();
```
Почему тут возвращается отдельный пустой ArrayList, а не тот, который возвращает repository?
Возможно, если есть необходимость гарантированно вернуть пустой список, то лучше использовать Collections.emptyList()?

efr21 commented on src/main/java/school/faang/user_service/service/event/EventParticipationService.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public Integer getParticipantsCount(Long eventId) {
+        if (eventId == null || eventId <= 0) {
```
Входные параметры лучше и безопаснее валидировать на уровне контроллера, не на уровне сервиса.

https://www.baeldung.com/spring-validate-requestparam-pathvariable

SoslanDzgoev committed:
Added EventParticipationController and UserMapper

- Implemented EventParticipationController with methods to register and unregister participants, and to get participants list and count
- Added UserMapper interface to map between User and UserDTO objects

SoslanDzgoev committed:
Added EventParticipationController and UserMapper

- Implemented EventParticipationController with methods to register and unregister participants, and to get participants list and count
- Added UserMapper interface to map between User and UserDTO objects
- Сorrected the mistakes.

SoslanDzgoev committed:
Added EventParticipationController and UserMapper

- Implemented EventParticipationController with methods to register and unregister participants, and to get participants list and count
- Added UserMapper interface to map between User and UserDTO objects
- Сorrected the mistakes.

eRqa commented on src/main/java/school/faang/user_service/controller/event/EventParticipationController.java:
```diff
@@ -0,0 +1,40 @@
+@RequiredArgsConstructor
+public class EventParticipationController {
+
+    private final EventParticipationServiceImpl eventParticipationServiceImpl;
```
Ты объявил интерфейс EventParticipantService и сделал его имплементацию EventParticipationServiceImpl. Здесь правильнее делать зависимость от **абстракции**, а не от **реализации**, то есть от EventParticipantService. Спринг сам найдет имплементацию EventParticipationServiceImpl и заинжектит

eRqa commented on src/main/java/school/faang/user_service/mapper/UserMapper.java:
```diff
@@ -0,0 +1,18 @@
+
+import java.util.List;
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
```
Все что в скобках не нужно писать
Так нужно писать если мы собираемся внутри какого то маппера, использовать какой то другой бин или маппер - например, если у нас есть несколько ДТОшек, в которых есть одно и тоже поле, и что бы в каждом маппере не писать одинаковый маппинг, можно это вынести в отдельный бин, и подключать его

Подробнее в пункте 5.2 https://www.baeldung.com/mapstruct

eRqa commented on src/main/java/school/faang/user_service/service/event/EventParticipationServiceImpl.java:
```diff
@@ -0,0 +1,54 @@
+
+        List<User> participants = eventParticipationRepository.findAllParticipantsByEventId(eventId);
+
+        if (participants.isEmpty()) {
```
Здесь получается если список пустой, возвращаем список participants, а если не пустой возвращаем participants
Здесь точно все правильно?)

eRqa commented on src/test/java/school/faang/user_service/service/event/EventParticipationServiceImplTest.java:
```diff
@@ -0,0 +1,106 @@
+
+@ExtendWith(MockitoExtension.class)
+class EventParticipationServiceImplTest {
+
```
Для метода getParticipants неплохо бы добавить тест когда eventId не корректный тогда должно выбросится исключение IllegalArgumentException

SoslanDzgoev committed:
Провел рефакторинг кода после ревью

SoslanDzgoev committed:
Провел рефакторинг кода после ревью
